### DISCUSSION TITLE
How to find the index of a given text string

### DISCUSSION THREAD

**Discussion by JoeRoddy** [ueberdosis/tiptap#3324] [General Questions] [answered]
Hi!  I'm building a feature where we provide automatic bookmarks for headings, so that you can click the bookmark and go jump your cursor straight to the end of that heading.

I'm wondering what the right way to achieve this is.  Currently, I'm searching the `editor.getText()` string to find the index of the header, and focusing that index:
```javascript
 const focusText = (text: string) => {
    const editorContent = editor?.getText();
    const index = editorContent?.indexOf(text) || -1;
    index >= 0 && editor?.commands.focus(index + text.length);
  };
```

The problem with this is that the index in `.getText()` is not the same as the index in the actual editor, once you add a few new lines.  Here's a video example of the problem:
https://user-images.githubusercontent.com/16690442/196482760-3a9cf470-b01d-463e-bb55-438d7b2dfba6.mov

Is there a better method for attempting to set the cursor to a given element?  Is it possible to just save the index position when I'm building out my list of headers?  I didn't see an api that provided the index position along with the given content.  Right now, here's how we're generating our list of headers:
```javascript
const headers =
    editor
      ?.getJSON()
      .content?.filter((c) => c.type === 'heading')
      .map((c) => c.content?.[0].text) || [];
```
Is there an api that will give me the markup, along with the real index of each html element?  

Thanks a ton!
-Joe

**Comment by sereneinserenade** [comment, marked as answer]
You can use [unique-id-extension](https://tiptap.dev/api/extensions/unique-id) to give a uuid to every heading and then you can focus the heading with given id like this.

```ts
const focusHeadingWithGivenUuid = (uuid: string) => {
  let posOfHeading = -1

  editor.state.doc.descendants((node, pos) => {
    if (node.type.name !== 'heading' || node.attrs.id !== uuid) return // edit `||` instead of `&&`

    posOfHeading = pos // `pos + node.nodeSize - 1` if you wanna focus at the end of line, maybe you don't need that `- 1` idk, you can try it out
  })

  if (posOfHeading !== -1) editor.commands.focus(posOfHeading)
  // or you can also use `setTextSelection` instead of `focus`
  // if (posOfHeading !== -1) editor.commands.setTextSelection(posOfHeading)
}
```

**Reply by JoeRoddy** [reply]
Awesome, thanks for the response @sereneinserenade !
